### PR TITLE
Support for Erase with --no-reset

### PR DESCRIPTION
**Pull request by VeliMattiLahtela**
This PR will bring --no-reset support for erase functionality.

Still missing changes for unit tests.

### CODE REVIEW

**Comment by RomanSaveljev** [comment]
I will try to use it in the daemon side

**Comment by RomanSaveljev** [comment]
Seems to work just fine with K64F

File: mbed_flasher/erase.py [hunk 1]
```diff
@@ -95,20 +112,26 @@ def erase_board(self, mount_point, serial_port):
...
-            sleep(SLEEP_TIME_FOR_REMOUNT)
-            success = self.reset_board(serial_port)
-            if isfile(join(mount_point, 'ERASE.ACT')):
-                success = EXIT_CODE_ERASE_FAILED_NOT_SUPPORTED
-            if success != 0:
-                self.logger.error("erase failed")
-                return success
+            t = Thread(target=self.runner, args=([mount_point, 'ERASE.ACT'],))
+            t.start()
+            if not no_reset:
```
Comment (RomanSaveljev) [inline mbed_flasher/erase.py]: So, the board will be reset even before we finished waiting for remount? Is this right?

After remount the MCU will be left in some kind of jammed state. I think it should be safe to reset it afterwards

File: mbed_flasher/erase.py [hunk 2]
```diff
@@ -95,20 +112,26 @@ def erase_board(self, mount_point, serial_port):
...
+                success = self.reset_board(serial_port)
+            while True:
```
Comment (RomanSaveljev) [inline mbed_flasher/erase.py]: This spin-locking is not a good idea IMO. Runner thread will sleep itself for 2 seconds, so we are going to be running at full CPU for at least 2 seconds. There is `.wait(timeout)` method. Is it not suitable here?

File: mbed_flasher/erase.py [hunk 3]
```diff
@@ -95,20 +112,26 @@ def erase_board(self, mount_point, serial_port):
...
+                if not t.is_alive():
+                    break
+            #if isfile(join(mount_point, 'ERASE.ACT')):
```
Comment (RomanSaveljev) [inline mbed_flasher/erase.py]: Commented code

File: mbed_flasher/main.py [hunk 4]
```diff
@@ -171,6 +171,7 @@ def argparser_setup(self, sysargs):
         # Initialize erase command
         parser_erase = get_resource_subparser(subparsers, 'erase', func=self.subcmd_erase_handler, help='Erase given resource')
         parser_erase.add_argument('--tid', '--target_id', help='Target to be erased or ALL, also multiple targets can be given. Short target_id matches boards by prefix', default=None, metavar='TARGET_ID', action='append')
+        parser_erase.add_argument('--no-reset', help='Do not reset device after erase', default=None, dest='no_reset', action='store_true')
```
Comment (RomanSaveljev) [inline mbed_flasher/main.py]: `default=False` for cleanliness?

**Comment by RomanSaveljev** [comment]
@mbed-ci integrate

File: mbed_flasher/erase.py [hunk 5]
```diff
@@ -80,8 +82,23 @@ def reset_board(self, serial_port):
                 return EXIT_CODE_SERIAL_RESET_FAILED
         port.close()
         return EXIT_CODE_SUCCESS
+
+    def runner(self, drive):
+        start_time = time()
+        while True:
+            sleep(0.3)
+            if platform.system() == 'Windows':
+                out = os.popen('dir %s' % drive[0]).read()
```
Comment (jupe) [inline mbed_flasher/erase.py]: can this raise exception?

File: mbed_flasher/erase.py [hunk 6]
```diff
@@ -95,20 +112,23 @@ def erase_board(self, mount_point, serial_port):
...
-            if success != 0:
-                self.logger.error("erase failed")
-                return success
+            t = Thread(target=self.runner, args=([mount_point, 'ERASE.ACT'],))
+            t.start()
+            while t.is_alive():
+                t.join(0.5)
+            if not no_reset:
+                success = self.reset_board(serial_port)
+            if not no_reset:
```
Comment (juhhov) [inline mbed_flasher/erase.py]: Wwould one "if not no_reset:" be enough?

File: mbed_flasher/erase.py (hunk 5 above)
Comment (RomanSaveljev) [inline mbed_flasher/erase.py]: This needs to be converted to use `subprocess` module, because `os.popen` has been deprecated

If it raises an exception, the thread will terminate. I think it is good, if it visibly shows that exception, because it should not happen (unless some bad system misconfiguration?)

File: mbed_flasher/erase.py [hunk 7]
```diff
@@ -27,9 +30,9 @@
 EXIT_CODE_PYOCD_ERASE_FAILED = 27
 EXIT_CODE_NONSUPPORTED_METHOD_FOR_ERASE = 29
 EXIT_CODE_IMPLEMENTATION_MISSING = 31
-SLEEP_TIME_FOR_REMOUNT = 5
 EXIT_CODE_ERASE_FAILED_NOT_SUPPORTED = 33
 EXIT_CODE_TARGET_ID_MISSING = 34
+ERASE_VERIFICATION_TIMEOUT = 100
```
Comment (RomanSaveljev) [inline mbed_flasher/erase.py]: Is this too long?

File: test/test_erase.py [hunk 8]
```diff
@@ -81,6 +81,20 @@ def test_erase_with_target_id(self, mock_stdout):
...
+        mbeds = mbed_lstools.create()
+        devices = mbeds.list_mbeds()
+        eraser = Erase()
+        ret = None
+        for item in devices:
+            if item['target_id']:
+                ret = eraser.erase(target_id=item['target_id'], method='simple', no_reset=True)
+                break
+        self.assertEqual(ret, 0)
+        time.sleep(4)
```
Comment (RomanSaveljev) [inline test/test_erase.py]: Why is this needed here?

**Comment by RomanSaveljev** [comment]
I want a release!
